src/blockstore/rocksdb.zig

```diff
@@ -0,0 +1,198 @@
+        }
+
+        // open rocksdb
+        const database: rocks.DB, //
+        const cfs: []const rocks.ColumnFamily //
+        = try callRocks(
+            logger,
+            rocks.DB.openCf,
+            .{
+                allocator,
+                path,
+                .{ .create_if_missing = true, .create_missing_column_families = true },
+                column_family_descriptions,
+            },
+        );
+        defer allocator.free(cfs);
+
+        // allocate handle slice
+        var cf_handles = try allocator.alloc(rocks.ColumnFamilyHandle, column_families.len);
+        errdefer allocator.free(cf_handles); // kept alive as a field
+
+        // initialize handle slice
+        for (1..cfs.len) |i| {
+            cf_handles[i - 1] = cfs[i].handle;
+        }
+
+        return .{
+            .allocator = allocator,
+            .db = database,
+            .logger = logger,
+            .cf_handles = cf_handles,
+        };
+    }
+
+    pub fn deinit(self: Self) void {
+        self.allocator.free(self.cf_handles);
+        self.db.deinit();
+    }
+
+    pub fn free(_: Self, bytes: []const u8) void {
+        rocks.free(bytes);
+    }
+
+    pub fn put(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        cf_index: usize,
+        key: cf.Key,
+        value: cf.Value,
+    ) !void {
+        const key_bytes = try cf.key().serializeToRef(self.allocator, key);
+        defer key_bytes.deinit();
+        const val_bytes = try cf.value().serializeToRef(self.allocator, value);
+        defer val_bytes.deinit();
+        return try callRocks(
+            self.logger,
+            rocks.DB.put,
+            .{ &self.db, self.cf_handles[cf_index], key_bytes.data, val_bytes.data },
+        );
+    }
+
+    pub fn get(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        cf_index: usize,
+        key: cf.Key,
+    ) !?cf.Value {
+        const val_bytes = try self.getBytes(cf, cf_index, key) orelse return null;
+        defer val_bytes.deinit();
+        return try cf.value().deserialize(cf.Value, self.allocator, val_bytes.data);
+    }
+
+    pub fn getBytes(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        cf_index: usize,
+        key: cf.Key,
+    ) !?BytesRef {
+        const key_bytes = try cf.key().serializeToRef(self.allocator, key);
+        defer key_bytes.deinit();
+        const val_bytes: rocks.Data = try callRocks(
+            self.logger,
+            rocks.DB.get,
+            .{ &self.db, self.cf_handles[cf_index], key_bytes.data },
+        ) orelse return null;
+        return .{
+            .allocator = val_bytes.allocator,
+            .data = val_bytes.data,
+        };
+    }
+
+    pub fn delete(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        cf_index: usize,
+        key: cf.Key,
+    ) !void {
+        const key_bytes = try cf.key().serializeToRef(self.allocator, key);
```

Comment:
theres a lot of abstract types (key: Value, Serializer, KeySerializer) which i get are to support different types of serialization/types of dbs but it gets extremely confusing very quickly - for an initial implementation we should only use these types of abstractions only when necessary

if we want to keep this - we should add more docs (both doc comments and more of a post explaning how the abstractions work)

Reply:
I thought the custom serializers were necessary because agave uses different serializers. Agave uses traits for this, but it's basically the same idea.

But actually bincode is able to do most of it just fine, so I switched to using the bincode serializer for almost everything. The only exceptions are shreds, which need to pass through untouched. There are also some cases where the endianness is big instead of little, but maybe I don't need to worry about that since we're not targeting ledger binary compatibility with agave.

I'll see if I can come up with a more concrete solution for serializers.